### DISCUSSION TITLE
WEB Prototype

### DISCUSSION THREAD

**Discussion by hrgdavor** [jscad/OpenJSCAD.org#944] [Show and tell]
This prototype is moving to another repo to allow another developer/s to contribute and to even release some official versions until it is ready to be contributed back to jscad. Repo is https://github.com/hrgdavor/jscadui

current work is in https://github.com/hrgdavor/jscadui/pull/9

it is about making actual usable components for users to combine jscad into an application, vanilla-js and a React example.

https://user-images.githubusercontent.com/2480762/206038051-31938795-2f32-44ef-b29b-b428bdef22d0.mp4

parameter form is working 2023-01-21

https://user-images.githubusercontent.com/2480762/213881066-e947e0ae-1b80-4a06-8234-fa9b1c974a19.mp4

right renderer works and orbit controls (both threejs) plus making a camera navigation gizmo
more details here: https://github.com/hrgdavor/jscadui/discussions/10

--------------------------------
I will change this text and also post replies so those interested can  follow the progress.

Disclaimer:
 - whatever code is good will be copied to final version, rest will be left behind :)

branch: https://github.com/jscad/OpenJSCAD.org/tree/proto-web2

Looking at the code
 - goal is to take advantage of tools available, so for example VSCode can provide as much autocomplete support
 - doing tweaks so it integrates nicely in VSCode

## checklist [2022-05-07]
copied from readme in the  prototype branch, and it will grow as more is clear what I plan to play with there

 - [x] index page that can be loaded by starting a web server
 - [x] make gui skeleton with few buttons
 - [x] integrate translations from jscad web
 - [x] initial layout (top buttons, editor panel, empty model panel)
 - [x] toggle editor
 - [x] toggle settigns
 - [x] store settings to localStorage
 - [x] integrate themes from jscad web (only for 3d view)
 - [ ] themes for UI (dark/light) 
 - [x] EDITOR codemirror 6
 - [x] EDITOR javascript syntax higlight
 - [x] ELECTRON: load same index page as electron app
 - [x] VIEWER - regl renderer 
 - [x] VIEWER - Three.js renderer 
 - [x] VIEWER - Babylon renderer 
 - [x] WORKER - run script
 - [x] VIEWER - parameters
 - [ ] WEB - file upload and changes syncing with worker
 - [ ] ALL - generalize file access
 - [ ] ELECTRON - allow working with local files
 - [ ] ELECTRON: recognize when in electron
 - [ ] ELECTRON: if in it, enable saving, not just reading script
 - [ ] DEMO-WEB: generate web page so proto is available for playing with as-is

| feature | regl | Threejs | Babylonjs |
|---| :-: | :-: | :-: |
| get/set camera | ✓ | ✓ | ✓ |
| get/set camera pan |  ✓ |   | ✓ |
| theme background | ✓ | ✓ | ✓ |
| line segments | ✓ | ✓ | ✓ |
| mesh | ✓ | ✓ | ✓  |

nice to have or to consider if worth doing

 - [ ] extract reusable OrbitControls to be used with different renderer (regl,three, babylon)

## progress

support for paramaters 

https://user-images.githubusercontent.com/2480762/167273430-72942aef-0718-4700-a6a8-ac2f128c0214.mp4

**Comment by hrgdavor** [comment]
- checklist is growing and layout is taking shape
 - codemirror 6 is being used
 - esm modules feel good to use, works nicely in vscode
 - not planning to use typescript, still nice completions are available in vscode with JS and esm

electron app just works, same as test web page 
![image](https://user-images.githubusercontent.com/2480762/140661148-28303a33-005b-4f42-8a13-4a0135a9fb0a.png)

later on when file handling is implemented electron app will need a bit special handling to take advantage of real file access.

**Comment by z3dev** [comment]
@hrgdavor how goes the new application?

I took a look at the code and the build. Looks really interesting and compact. I’ll continue to keep watch so let me know if there are any questions. Also, if there are some nice improvements then let me know and I’ll try things out.

Sadly, I’m back at full time work again, and don’t have a lot of time for JSCAD. So, please be patient.

**Reply by hrgdavor** [reply]
@z3dev  Hi, I am glad you are following the branch. I was mostly playing with jsx and what  I can do with plain js without plugins so the superfast esbuild can be used. Also looking to what extent can the JS code there be friendly to autocomplete in vscode.

On top of that there are few other thing I am discussing in discord channel to bring into jscad on discord.
 - OCCT
 - another CSG engine (to test if it can be better than csg.js in some aspects)
 - SDF (signed distance fields) ... maybe maybe in some way

I have a tendency to go sideways like that, but the main reason development was slow last few weeks is project deadlines ... end of the year is nearing and invoices need to go out before holidays, and stuff finished before that :).

also check a hot new drawing and diagramming app: https://www.tldraw.com

good luck on your new job, and may it be such job that u do not get burned-out, and have time for jscad :)

**Comment by hrgdavor** [comment]
@z3dev  mini update: I have been playing more with my ui lib, but finally started to assemble code for new worker.
I am testing drag and drop, an forwarding the dropped files to worker.

This is a tricky step to get right as I want also to support same file access in electron app. this will take some time, so I will tackle it in multiple iterations while working on execution and render for single script that is edited in the embeded editor.

**Reply by z3dev** [reply]
Definitely tricky so take your time.

There are probably a few design decisions to make as the automatic reload of files will allow users to use an external IDE or editor to make changes. The ability to use a debugger in the electron application is also desirable.

**Reply by hrgdavor** [reply]
The dev tools that you have in chrome is also shipped with electron apps by default. The F12 keyboard shortcut is reserved, but there is an alternative keyboard shotcut: CTRL+SHIFT+i

![image](https://user-images.githubusercontent.com/2480762/144810133-688c3873-b0d9-40ad-9696-feec90185974.png)

Fun fact, you can get dev tools for Discord app and Github desktop (they are obviously also electron apps)

**Reply by z3dev** [reply]
So,is VSCODE from Microsoft. Ha ha ha!

**Reply by hrgdavor** [reply]
Yes, Atom was also an attempt to better than sublime that I was following, but was not impressed.

I am not fan of MS at all, but I am impressed how well vscode works and is made in JS as electron app.

They also have an online version of vscode...

 I will likely look into supporting vscode as inline editor for jscad web as alternative for codemirror ... it is worth exploring. 

Currently I am finding it pretty useful while making the prototype. As mentioned here more than few times, I still have no plans to switch to Typescript, but I am keen to get any benefits of code completion in vscode with plain JS. Pressing a shortcut to jump to a function declaration is fun :)

**Comment by hrgdavor** [comment]
it has been slow 2 months, still here is something to look forward to:
https://doc.babylonjs.com/divingDeeper/materials/advanced/transparent_rendering#order-independent-transparency

threejs and babylon have some level of support for better transparency rendering.
so supporting all 3 engines gives choice to use what fits purpose for given case

**Comment by hrgdavor** [comment]
small progress update. Finally finding some time to continue working on the prototype.
 - regl renderer implemented with orbit controls
 - showAxes option integrated vith viewer
 - showGrid option integrated with viewer
![image](https://user-images.githubusercontent.com/2480762/151123395-5bf793e0-3948-4417-97a7-e8b7d32743a4.png)
![image](https://user-images.githubusercontent.com/2480762/151123474-3b34c4aa-4589-493f-94ac-b3b0c04b5dc0.png)

it is not much, but finally moving fwd :)

**Comment by hrgdavor** [comment]
new progress :) @z3dev ;)
 - inital implementation of regl renderer and threejs renderer 
 - switching renderer live
 - works in web and electron

https://user-images.githubusercontent.com/2480762/152426577-7170dc11-99e1-4c13-a363-449cf2c8ba39.mp4

**Comment by hrgdavor** [comment]
bybylon.js basic scene with grid and axes ... needs more tweaking, but it works.. babylon view  is currently 6.7 mb , but it should go down in size when customized closer to 1MB..

https://user-images.githubusercontent.com/2480762/152607806-0d244616-f0b5-4212-8db5-691e5d5ec0b7.mp4

**Comment by hrgdavor** [comment]
after setting up threejs and regl basic scene with grid and axis, babylon is taking some time to tune to match camera view

One major difference is handiness
https://forum.babylonjs.com/t/unexpected-behavior-for-z-up-right-handed-coordinate-system/1090/7

also default fov in babylon is 0.8 which is not exact match to regl and threejs default, but easy fix `camera.fov = Math.PI / 4`

after playing around the views are getting closer

https://user-images.githubusercontent.com/2480762/155024859-a311cc12-6b19-472b-831d-3f917804c29b.mp4

**Reply by z3dev** [reply]
Now why would Babylon go with left handed system, and Y up? That is just lame. They gave no reason for that decision other then saying… “Maya did it”. The users are correct… it’s a mind F

**Comment by hrgdavor** [comment]
I was trying to use library specific code to render grid and axes, but both threejs ang babylonjs implementations are missing things that would enable me to make all 3 look the same.

The most robust way to get grid and axis to work and look the same it to generate properly colored line segments. I will switch to model where jscad(web ui) will generate axis lines and grid and give it along with models from executed jscad script, 

I will try make it so all 3 engine wrappers support grouping of objects and multiple scenes to be rendered on a section of the screen.

this will also mean no need to reimplement things in wrappers if we add support to different grid types in jscad. 

Threejs and sharing of objects:
>You can share geometry along different Scenes.
>You can't share meshes along different Scenes.
>You can't share geometry/meshes/scenes along different Renderers (yet).

**Comment by hrgdavor** [comment]
Code that used grid and axis for the specific library (regl,threejs,babylon) was removed and implemented a manually generated grid and axes

 Even though opacity in jscad theme is 1 for color1 and 0.5 for color2, the closest match to the color on the website
 is 0.2 and 0.1
![image](https://user-images.githubusercontent.com/2480762/159188726-66a12345-75d4-48c1-92dd-f86dc8918f36.png)

this is also the start of supporting line-segments geometry type in renderer (currently only threejs viewer)

 Opacity per line-segment is tricky and was added only in threejs 127. so splitting the grid into 2 separate set of lines
 allows to have different opacity for grid and subgrid via opacity attribute and keep rgb colors. Also this removes the need to use
 color per line-segment (no need for an additional buffer with colors).

It remains to be seen if using bunch of lines is a good choice for grid (if it has some perf penalty)

This is also first step towards defining common format for geometries (that uses TypedArrays).  https://github.com/jscad/OpenJSCAD.org/blob/proto-web2/packages/web2/geom.format.md

**Comment by hrgdavor** [comment]
Nothing visual to report as I have recreated succesfully grid and axis in babylon and screen above is representative, but soon the grid and axis will be consistent for all 3 engines. This also will open posibility to allow for more types of grids or axis visualization.

there are 2 major geometries to display:
 - mesh
 - lines

lines now work in Threejs and babylon but some interesting implementation details came-up.

The minor issue is with a use-case that not used in jscad yet: color per vertex for lines but will likely be present in mesh too. 
It seems like colors data buffer uses 4 channels regardless if alpha is used or not. this means worker will need this type of information when producing buffers. Also this info should be included in output. I have plans to test all 3 engines in parallel at some point, so code needs to be aware if there is need to convert color buffer from/to 3 o4 channels. 

Looking at the code, it is definitely 4 channel in babylon.
Threejs changes channels per color depending vertexAlpha.
Regl, I need to see what is or not supported in the shader.

**Reply by z3dev** [reply]
Colors are convert to RGBA so everything should be fine.

**Reply by hrgdavor** [reply]
> Colors are convert to RGBA so everything should be fine.

Yes, it is not a problem really.
I am just looking to allow for a optimized path so that buffers created by worker can be used without additional convert step. 
There are also some differences I think how each handles when general color is used in combo with color per vertex.

**Comment by hrgdavor** [comment]
ok, this is too freaky so I have to write it down.

I used lines to draw the grid based on the theme, and was getting strange result. babylonjs and threejs were blending the semi transparent grid lines correctly but regl looked different and visually not correct. ... after trying all sorts of things, and inspecting the different it looked like regl was blending twice, once towards white and once more correctly .... after smashing my head on the keyboard for some times the solution that seems to work is:

```js
canvas.style.background = 'black'
```

I hope this really fixes the issue. 

the color `[0,0,0,0.5]` should not blend like this to the brown background
![image](https://user-images.githubusercontent.com/2480762/162638035-f2c57d23-345c-47f1-8939-fbb7ac8111aa.png)

it should be like this (and it is with the above mentioned trick)
![image](https://user-images.githubusercontent.com/2480762/162638053-710681cb-7142-4cbf-a0a3-a88c5a17609f.png)

**Comment by hrgdavor** [comment]
finally one further step was made:
 - support for lines (and alpha)
 - support for multicolored lines (will need to make a PR for jscad, as it is not properly supported on master)

video shows 
 - theme switching, 
 - switching render engines without noticable visual change (that is a good thing :) )


https://user-images.githubusercontent.com/2480762/162638934-7a945625-420d-4041-98fc-5735fcc34c6b.mp4

**Comment by hrgdavor** [comment]
I have just made another step forward: mesh support (but with some quirks ATM)
I have some lighting to fix, and also babylonjs for some weird reason uses left handed system (this adds a bit of complexity to the worker to support it properly) https://forum.babylonjs.com/t/vertexdata-computenormals-in-right-handed-system/25195

Also when switching viewers, if they are different orientation currently I need to run the script again manually to get properly oriented triangles

https://user-images.githubusercontent.com/2480762/163730190-d252c9ec-a329-4a95-8a5b-b6d7d6e2c6c6.mp4

**Comment by hrgdavor** [comment]
after 20 days, another small update

support for paramaters is integrated (there will be UX changes) and work for the purpose of testing worker communication and parameter updating

this also demonstrates how nice it is when worker is not killed, but reused. fast scripts have an option to give instant feedback on param changes and I think this will be also a nice incentive to make faster scripts.

https://user-images.githubusercontent.com/2480762/167273430-72942aef-0718-4700-a6a8-ac2f128c0214.mp4

**Comment by hrgdavor** [comment]
Last missing piece is closer to be ready: Reusable orgit contros, and a bonus a nice camera gizmo.

https://user-images.githubusercontent.com/2480762/205169462-dc4d3be3-d58f-4ed0-b07b-89249412945d.mp4

I have gathered a lot of experience with Monaco, and runnig code with imports in workers. I am able to run code taht usese `require` but also new ES modules `imports`, and it is done by convertin ES modules via babel into code with `require`.

**Comment by hrgdavor** [comment]
Themes in all 3 engines

https://user-images.githubusercontent.com/2480762/206038051-31938795-2f32-44ef-b29b-b428bdef22d0.mp4

**Comment by hrgdavor** [comment]
some more progress with prototype. Loading of folder projects works, it can handle ESM modules, cjs scripts (require) and even typescript projects. Also worker improvements that allow script developers to optimize execution when parameters change and cache expensive operations.


current jscad

https://user-images.githubusercontent.com/2480762/213881095-dcd72bce-881a-463d-a105-c9ad9c0b44bb.mp4

my new prototype is still ugly, but things are progressing bit by bit

It was my mission from the first attempt at making a new jscad prototype to allow script developers to optimize  for performance and bling.

in this example I took a complex model and added caching so that model is kept in case only View Options parameters change. Those only affect rotation of the parts.

also I enhanced parameter definition to allow for live updates for any of the parameter (it is not default, but opt-in)


https://user-images.githubusercontent.com/2480762/213881066-e947e0ae-1b80-4a06-8234-fa9b1c974a19.mp4

**Comment by z3dev** [comment]
@hrgdavor have you seen this? Wow!

https://firepad.io/

or this!

https://etherpad.org/

**Reply by hrgdavor** [reply]
@z3dev it is great to see there is not one but at least two tools like that.

 I personally do not have a use-case where I would collaborate on jscad code in this way, but will be willing to assist if someone wants to to set it up. 

Thanks for the links, both are very useful in general and it is good to know they exist in case I need a setup like that.
